hashes/src/sha256.rs

```diff
@@ -989,7 +1012,7 @@ mod tests {
             0x8f, 0xf1, 0xf7, 0xa9, 0xd5, 0x69, 0x09, 0x59,
         ];
         let midstate_engine =
-            sha256::HashEngine::from_midstate(sha256::Midstate::from_byte_array(MIDSTATE), 64);
+            sha256::HashEngine::from_midstate(sha256::Midstate::new(MIDSTATE, 64));
```

Comment:
I'm skeptical about trying to hyperoptimize codegen in unit tests. In practice I do not believe there would be a measurable difference in code size (unless the `Midstate::new` function could be compiled out entirely which requires 100% adherence in all dependencies, ok) nor in performance (where we have a branch that can be predicted 100% of the time since it's checking whether a constant value has some low bits clear).

But yes, in example code we ought to use a `static` or `const` or whatever, to emphasize that you almost certainly want to be calling this function with a fixed pre-vetted constant and not trying to be clever.

Reply:
That's not really the point. If it's the best way to do so in examples it's probably the best in tests as well. If not for any other reason, at least to not give readers bad ideas.

The current doc doesn't have an example demonstrating proper use so people may look into the code and find this.

Reply:
> If it's the best way to do so in examples it's probably the best in tests as well.

This is definitely not true in general. The point of tests are to find edge cases and abuse the API to the extent allowable (and maybe beyond that).

In this case, sure, we can make the test a little noisier if it makes you happy. But I don't believe that anybody who somehow finds this type without knowing what they're doing, reads through the "don't use this type unless you know what you're doing" docs, then goes digging through unit tests to find "example usage", will benefit from us trying to microoptimize their code by proxy.

Reply:
Fair point given this is niche. I would like to avoid us making it a habit since I found that better code also makes refactoring tests easier (typically not using `vec!` like many people love to do, which is unneeded in maybe 99% of cases). I'd still really like having an example in the doc.

Not something I would hold a PR for (I would've clicked "request changes" otherwise).

Reply:
I added a patch using `static` in one of the unit tests.